cmd/deploy.go

```diff
@@ -127,12 +138,93 @@ func runDeploy(cmd *cobra.Command, _ []string) (err error) {
 		}
 	}
 
-	return client.Deploy(config.Path)
+	return client.Deploy(cmd.Context(), config.Path)
 
 	// NOTE: Namespace is optional, default is that used by k8s client
 	// (for example kubectl usually uses ~/.kube/config)
 }
 
+func credentialsProvider(ctx context.Context, registry string) (docker.Credentials, error) {
+
+	result := docker.Credentials{}
+	credentials, err := config.GetCredentials(nil, registry)
+	if err != nil {
+		return result, errors.Wrap(err, "failed to get credentials")
+	}
+
+	if credentials != (containersTypes.DockerAuthConfig{}) {
+		result.Username, result.Password = credentials.Username, credentials.Password
+		return result, nil
+	}
+
+	fmt.Print("Username: ")
+	username, err := getUserName(ctx)
+	if err != nil {
+		return result, err
+	}
+
+	fmt.Print("Password: ")
+	bytePassword, err := getPassword(ctx)
+	if err != nil {
+		return result, err
+	}
+	password := string(bytePassword)
+
+	result.Username, result.Password = username, password
+
+	return result, nil
+}
+
+func getPassword(ctx context.Context) ([]byte, error) {
```

Comment:
getPassword, getUserName -- I needed to make this in order to make those operations cancellable. Usually blocking read can be cancelled by closing the reader. For some reason this is not working with `os.Stdin`.

Reply:
It's nice to see the logic for interacting with the user where it belongs, here in `cmd` rather than the client library!

Usually this context creation and channel os.Signal listeners are something that is done at the very root of the process.  While I am not extremely familiar with Cobra, I would assume that this is one of its features, and we may want to change to using the functions in its API which include the context object.  Perhaps that will help.

In particular, I am thinking we may need to update to executing the command using [`ExecuteContext`](https://pkg.go.dev/github.com/spf13/cobra?utm_source=godoc#Command.ExecuteContext).  Perhaps by simply making this change, the command will be run using a cancelable context under the hood, and respond to events correctly?

Reply:
>In particular, I am thinking we may need to update to executing the command using ExecuteContext. 

We are already doing that. Now the context should be available via `cmd.Context()`.

Reply:
see: https://github.com/boson-project/func/blob/main/cmd/run.go#L69